### PR TITLE
Move id converter inside of router phase4

### PR DESCRIPTION
**Pull request by SophieGuo410**

### CODE REVIEW

File: ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3MultipartCompleteUploadHandler.java [hunk 1]
```diff
@@ -258,7 +258,7 @@ private Callback<String> idConverterCallback(BlobInfo blobInfo, String blobId) {
         String serviceId = blobInfo.getBlobProperties().getServiceId();
         retryExecutor.runWithRetries(retryPolicy,
             callback -> router.updateBlobTtl(blobId, serviceId, Utils.Infinite_Time, callback,
-                QuotaUtils.buildQuotaChargeCallback(restRequest, quotaManager, false)), this::isRetriable,
+                QuotaUtils.buildQuotaChargeCallback(restRequest, quotaManager, false), null, null), this::isRetriable,
```
Comment (Arun-LinkedIn) [inline ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3MultipartCompleteUploadHandler.java:261]: Just a note. I think we may have to move id conversion inside Router here too later (when we enable dual writes for Multipart uploads).

File: ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouter.java [hunk 2]
```diff
@@ -551,38 +553,43 @@ public Future<Void> undeleteBlob(String blobId, String serviceId, Callback<Void>
...
+   *                            blob permanent
+   * @param callback            The {@link Callback} which will be invoked on the completion of a request.
+   * @param restRequest         The {@link RestRequest} of updateBlobTtl
+   * @param idConverterCallback The id converter callback which take the String type instead of void.
    * @return A future that would contain information about whether the update succeeded or not, eventually.
    */
   @Override
   public Future<Void> updateBlobTtl(String blobId, String serviceId, long expiresAtMs, Callback<Void> callback,
-      QuotaChargeCallback quotaChargeCallback) {
+      QuotaChargeCallback quotaChargeCallback, RestRequest restRequest, Callback<String> idConverterCallback) {
```
[the remark this answers is not in the document]
Comment (SophieGuo410) [reply ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouter.java:568]: They are using the different type. the idConverter.convert can only take String type. And it's hard to use the generic type for convert method cause it touch lots of existing code IIUC.

File: ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3MultipartCompleteUploadHandler.java (hunk 1 above)
Comment (SophieGuo410) [reply ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3MultipartCompleteUploadHandler.java:261]: Yeah, we need to remove id convert logic in all router APIs later.

File: ambry-frontend/src/main/java/com/github/ambry/frontend/NamedBlobPutHandler.java [hunk 3]
```diff
@@ -248,11 +247,12 @@ private Callback<Void> securityPostProcessRequestCallback(BlobInfo blobInfo) {
...
     private Callback<String> routerPutBlobCallback(BlobInfo blobInfo) {
-      return buildCallback(frontendMetrics.putRouterPutBlobMetrics, blobId -> {
-        restResponseChannel.setHeader(RestUtils.Headers.BLOB_SIZE, restRequest.getBlobBytesReceived());
-        blobInfo.getBlobProperties().setBlobSize(restRequest.getBlobBytesReceived());
-        idConverter.convert(restRequest, blobId, blobInfo.getBlobProperties(), idConverterCallback(blobInfo, blobId));
-      }, uri, LOGGER, deleteDatasetCallback);
+      return buildCallback(frontendMetrics.putRouterPutBlobMetrics,
+          blobId -> {
+            restResponseChannel.setHeader(RestUtils.Headers.BLOB_SIZE, restRequest.getBlobBytesReceived());
+            idConverterCallbackForPut(blobInfo).onCompletion(blobId, null);
```
[the remark this answers is not in the document]
Comment (SophieGuo410) [reply ambry-frontend/src/main/java/com/github/ambry/frontend/NamedBlobPutHandler.java]: Updated.

File: ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouter.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (SophieGuo410) [reply ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouter.java]: After discussed with @justinlin-linkedin , I added a string call back to resolve it.

**Comment by justinlin-linkedin** [comment]
At this moment, the id converter transition for named blob is not yet completed. stitchblob doesn't have id converter. updateTtl should take in a blob name instead of blob id in when it's named blobs. We should probably do both of them for NamedBlobPutHandler in the same PR.

**Review by justinlin-linkedin — APPROVED** [review]

**Review by Arun-LinkedIn — APPROVED** [review]

**Pull request merged**